### PR TITLE
Add info log showing stats about data upon exiting stage_version context

### LINKED ISSUE #166: Log statistics at exit of stage_version context
I think it would be helpful if we could log some statistics after writing each version. So probably at the end of a `stage_version` context?
I would be interested in some information like
```
dataset_name: old_shape -> new_shape, chunk size, number of new raw chunks written, number of chunks reused
```
for every dataset impacted in the transaction.

### PR DESCRIPTION
**Pull request by peytondmurray**
This PR logs a message when exiting the `VersionedHDF5File.stage_version` context manager displaying some useful information about the data written. Information is displayed at the `INFO` log level.

This PR closes #166.

Here's a simple example taken from a test:

```python
import logging
import pathlib

import h5py
import numpy as np
from versioned_hdf5 import VersionedHDF5File

logging.basicConfig(level=logging.INFO)

dt = np.dtype('double')

# Clean out old data files
for p in pathlib.Path().glob("*.h5"):
    p.unlink()

# Create new versioned file
with h5py.File('foo.h5', 'w') as f:
    vf = VersionedHDF5File(f)
    with vf.stage_version('0') as sv:
        sv.create_dataset('bar', (2, 15220, 2),
                          chunks=(300, 100, 2),
                          dtype=dt, data=np.full((2, 15220, 2), 0, dtype=dt))
        sv.create_dataset('baz', (1, 10, 2),
                          chunks=(600, 2, 4),
                          dtype=dt, data=np.full((1, 10, 2), 0, dtype=dt))

with h5py.File('foo.h5', 'r+') as f:
    vf = VersionedHDF5File(f)
    with vf.stage_version('1') as sv:
        bar = sv['bar']
        bar.resize((3, 15222, 2))
        baz = sv['baz']
        baz.resize((1, (4) * 10, 2))
        baz[:, -10:, :] = np.full((1, 10, 2), 3, dtype=dt)
```

Running this produces displays the following:

```
INFO:versioned_hdf5.api:
  bar: Shape: None -> (2, 15220, 2); Chunks: None -> (300, 100, 2); New chunks written: 2; Number of chunks reused: 0
  baz: Shape: None -> (1, 10, 2); Chunks: None -> (600, 2, 4); New chunks written: 1; Number of chunks reused: 0
INFO:versioned_hdf5.api:
  bar: Shape: (2, 15220, 2) -> (3, 15222, 2); Chunks: ChunkSize((300, 100, 2)) -> ChunkSize((300, 100, 2)); New chunks written: 0; Number of chunks reused: 4
  baz: Shape: (1, 10, 2) -> (1, 40, 2); Chunks: ChunkSize((600, 2, 4)) -> ChunkSize((600, 2, 4)); New chunks written: 0; Number of chunks reused: 2
```

@ArvidJB @quarl What do you think of this? As long as this looks good to you, I'll move this out of draft and mark it ready for review.

### CODE REVIEW

File: versioned_hdf5/api.py [hunk 1]
```diff
@@ -230,3 +240,69 @@ def __repr__(self):
         else:
             return f"<VersionedHDF5File object \"{self.f.filename}\" (mode" \
                    f" {self.f.mode})>"
+
+    def get_hashes(self, name: str) -> Set[bytes]:
```
Comment (ArvidJB) [inline versioned_hdf5/api.py:244]: Does this need to be public?

File: versioned_hdf5/api.py [hunk 2]
```diff
@@ -230,3 +240,69 @@ def __repr__(self):
...
+            chunks_reused = len(old_hashes & new_hashes)
+            new_chunks_written = len(new_hashes - old_hashes)
+
+            msg.append(
+                f"  {name}: Shape: {old_shape} -> {new_shape}; "
+                f"Chunks: {old_chunks} -> {new_chunks}; "
+                f"New chunks written: {new_chunks_written}; "
+                f"Number of chunks reused: {chunks_reused}"
+            )
+        logger.info("\n".join(msg))
```
Comment (ArvidJB) [inline versioned_hdf5/api.py:308]: This should log at debug level by default, I think. Most of our production jobs run with info level debugging and would end up printing this, which our end users probably do not care about.

**Comment by ArvidJB** [comment]
Looks good to me! Just make sure the logging is at debug level, not info.

File: versioned_hdf5/api.py (hunk 1 above)
Comment (peytondmurray) [reply versioned_hdf5/api.py:244]: Definitely not, I'll change it now.

File: versioned_hdf5/api.py (hunk 2 above)
Comment (peytondmurray) [reply versioned_hdf5/api.py:308]: Sounds good to me.

**Comment by peytondmurray** [comment]
@asmeurer Before merging this I thought I'd ask you about the implementation of `__bool__` here, since it is user facing. Previously doing `bool(dataset)` would fail if the dataset was a scalar array because it would fall back to checking `len(dataset)`. With this change, scalar datasets will no longer throw errors. Thoughts on this behavior?

**Comment by asmeurer** [comment]
As I mentioned on Slack, the behavior should mirror h5py, which itself tries to mirror numpy. `bool(0_d_array)` works in numpy, and I haven't checked but it ought to work in h5py, so this change sounds correct (len(0_d) should fail, but I think len is just used as the default implementation of `bool`).

**Pull request merged**